### PR TITLE
Fix custom controlled versions of `SelectOracle`s and `QubitizationWalkOperator`s

### LINKED ISSUE #886: Should not override `.controlled` for custom controlled implementations
Bloq authors should always override `get_ctrl_system` and not `controlled`. I think only the old cirq-ft bloqs ( `QubitizationWalkOperator` and related) override `controlled` directly, which must be upgraded.

```py
from qualtran import Bloq
from qualtran.bloqs.for_testing.random_select_and_prepare import RandomPrepareOracle
from qualtran.bloqs.reflection_using_prepare import ReflectionUsingPrepare
from qualtran.drawing import show_bloq

prepare = RandomPrepareOracle.create(3, random_state=42)
R = ReflectionUsingPrepare(prepare)

class Wrapped(Bloq):
    @property
    def signature(self) -> 'Signature':
        return R.signature

    def build_composite_bloq(self, bb: 'BloqBuilder', **soqs: 'SoquetT') -> dict[str, 'SoquetT']:
        return bb.add_d(R, **soqs)

cb_direct = R.controlled().decompose_bloq()
cb_wrapped = Wrapped().controlled().decompose_bloq().flatten_once(lambda _: True)
```

```py
show_bloq(cb_direct)
```
direct

```py
show_bloq(cb_wrapped)
```
wrapped

More detailed notebook with above code: https://github.com/anurudhp/Qualtran/blob/2024-04-issue-controlled/controlled_issue.ipynb

### PR DESCRIPTION
**Pull request by anurudhp**
On top of #903 

fixes #886

### CODE REVIEW

File: qualtran/bloqs/hubbard_model.py [hunk 1]
```diff
@@ -219,28 +219,12 @@ def decompose_from_registers(
...
-            and self.control_val is None
-        ):
-            return SelectHubbard(self.x_dim, self.y_dim, control_val=control_values[0])
-        raise NotImplementedError(
-            f'Cannot create a controlled version of {self} with control_values={control_values}.'
-        )
+    def get_ctrl_system(
+        self, ctrl_spec: Optional['CtrlSpec'] = None
+    ) -> Tuple['Bloq', 'AddControlledT']:
+        from qualtran._infra.gate_with_registers import get_ctrl_system_for_single_qubit_controlled
```
Comment (fdmalone) [inline qualtran/bloqs/hubbard_model.py:225]: Why is the import happening inside?

Comment (anurudhp) [reply qualtran/bloqs/hubbard_model.py:225]: ah, this should be at top level.

I haven't yet decided if this is the proper way to implement this. Somehow the `control_val` interface is a bit ad-hoc and can't be type-checked properly. I'm open to suggestions for this interface.

File: qualtran/_infra/gate_with_registers.py [hunk 2]
```diff
@@ -549,3 +546,51 @@ def _circuit_diagram_info_(self, args: cirq.CircuitDiagramInfoArgs) -> cirq.Circ
...
+def get_ctrl_system_for_single_qubit_controlled(
+    bloq: GateWithRegisters, ctrl_spec: Optional['CtrlSpec'] = None
+) -> Tuple['Bloq', 'AddControlledT']:
+    """Wrapper to correctly wire a custom implementation of a single-qubit controlled version of a gate.
+
+    For any gate that supports a single-qubit controlled version
+    with an API that has `control_val` and `control_registers`.
+    Add the `get_ctrl_system` implementation as shown below:
+
+    .. code:: python
```
Comment (mpharrigan) [inline qualtran/_infra/gate_with_registers.py:560]: we actually render all this as markdown, so this directive won't work properly. I think the indented section should be rendered as code; you might need backticks if you want to say "python"

File: qualtran/_infra/gate_with_registers.py [hunk 3]
```diff
@@ -549,3 +546,51 @@ def _circuit_diagram_info_(self, args: cirq.CircuitDiagramInfoArgs) -> cirq.Circ
...
         return cirq.CircuitDiagramInfo(wire_symbols=wire_symbols)
+
+
+def get_ctrl_system_for_single_qubit_controlled(
+    bloq: GateWithRegisters, ctrl_spec: Optional['CtrlSpec'] = None
+) -> Tuple['Bloq', 'AddControlledT']:
+    """Wrapper to correctly wire a custom implementation of a single-qubit controlled version of a gate.
+
+    For any gate that supports a single-qubit controlled version
+    with an API that has `control_val` and `control_registers`.
```
Comment (mpharrigan) [inline qualtran/_infra/gate_with_registers.py:557]: can you be more specific than "with an api that has..."

I think you require that the bloq has attributes named `control_val` and `control_registers`, right? Does this need to be a `GateWithRegisters` like the annotation?

File: qualtran/_infra/gate_with_registers.py [hunk 4]
```diff
@@ -549,3 +546,51 @@ def _circuit_diagram_info_(self, args: cirq.CircuitDiagramInfoArgs) -> cirq.Circ
...
+            bb: 'BloqBuilder', ctrl_soqs: Sequence['SoquetT'], in_soqs: dict[str, 'SoquetT']
+        ) -> tuple[Iterable['SoquetT'], Iterable['SoquetT']]:
+            soqs = {ctrl_reg.name: ctrl_soqs[0]} | in_soqs
+            soqs = bb.add_d(cbloq, **soqs)
+            ctrl_soqs = [soqs.pop(ctrl_reg.name)]
+            return ctrl_soqs, soqs.values()
+
+        return cbloq, adder
+
+    raise NotImplementedError(f'Cannot create a controlled version of {bloq} with {ctrl_spec=}.')
```
Comment (mpharrigan) [inline qualtran/_infra/gate_with_registers.py:596]: for this to properly implement the `get_ctrl_system` method, it should delegate to the base class method if no specific construction is available. (This base method will wrap the bloq in `Controlled` metabloq).

File: qualtran/_infra/gate_with_registers.py [hunk 5]
```diff
@@ -549,3 +546,51 @@ def _circuit_diagram_info_(self, args: cirq.CircuitDiagramInfoArgs) -> cirq.Circ
...
+    """
+    if ctrl_spec is None:
+        ctrl_spec = CtrlSpec()
+
+    assert hasattr(bloq, 'control_val')
+    assert hasattr(bloq, 'control_registers')
+
+    if bloq.control_val is None and ctrl_spec.shapes in [((),), ((1,),)]:
+        cbloq = attrs.evolve(bloq, control_val=int(ctrl_spec.cvs[0].item()))
+        (ctrl_reg,) = cbloq.control_registers
```
Comment (mpharrigan) [inline qualtran/_infra/gate_with_registers.py:584]: so this assumes that `control_registers` is actually a property that depends on the `control_val` attribute, right? please document and/or code this more defensively

File: qualtran/bloqs/mean_estimation/mean_estimation_operator.py [hunk 6]
```diff
@@ -126,36 +125,36 @@ def decompose_from_registers(
...
+
+            def adder(
+                bb: 'BloqBuilder', ctrl_soqs: Sequence['SoquetT'], in_soqs: dict[str, 'SoquetT']
+            ) -> tuple[Iterable['SoquetT'], Iterable['SoquetT']]:
+                soqs = {ctrl_reg.name: ctrl_soqs[0]} | in_soqs
+                soqs = bb.add_d(cbloq, **soqs)
+                ctrl_soqs = [soqs.pop(ctrl_reg.name)]
+                return ctrl_soqs, soqs.values()
+
+            return cbloq, adder
```
Comment (mpharrigan) [inline qualtran/bloqs/mean_estimation/mean_estimation_operator.py:149]: you should probably make it so it's possible to factor this out as well, right? we should probably pick what the `control_val`/`cv` argument is called

**Review by mpharrigan — COMMENTED** [review]
This is a good idea. The current PR needs a little polish

File: qualtran/_infra/gate_with_registers.py [hunk 7]
```diff
@@ -549,3 +546,51 @@ def _circuit_diagram_info_(self, args: cirq.CircuitDiagramInfoArgs) -> cirq.Circ
...
+            ) -> Tuple['Bloq', 'AddControlledT']:
+                from qualtran._infra.gate_with_registers import get_ctrl_system_for_single_qubit_controlled
+
+                return get_ctrl_system_for_single_qubit_controlled(self, ctrl_spec)
+    """
+    if ctrl_spec is None:
+        ctrl_spec = CtrlSpec()
+
+    assert hasattr(bloq, 'control_val')
+    assert hasattr(bloq, 'control_registers')
```
Comment (mpharrigan) [inline qualtran/_infra/gate_with_registers.py:580]: I don't know if this is a better or worse idea, but: did you consider making this a mixin?

Comment (anurudhp) [reply qualtran/_infra/gate_with_registers.py]: this does make the code cleaner, and enables type-checking. I made a class `SpecializedSingleQubitControlledGate` (similar to the design of `UnaryIterationGate`), please have a look

File: qualtran/bloqs/mean_estimation/mean_estimation_operator.py (hunk 6 above)
Comment (anurudhp) [reply qualtran/bloqs/mean_estimation/mean_estimation_operator.py]: fixed this now, uses the same code as the others

File: qualtran/_infra/gate_with_registers.py (hunk 4 above)
Comment (anurudhp) [reply qualtran/_infra/gate_with_registers.py]: fixed

File: qualtran/_infra/gate_with_registers.py (hunk 3 above)
Comment (anurudhp) [reply qualtran/_infra/gate_with_registers.py]: updated the design to use a mixin class

File: qualtran/_infra/gate_with_registers.py (hunk 2 above)
Comment (anurudhp) [reply qualtran/_infra/gate_with_registers.py]: oops, fixed now

File: qualtran/_infra/gate_with_registers.py (hunk 5 above)
Comment (anurudhp) [reply qualtran/_infra/gate_with_registers.py]: added a check that raises TypeError

**Review by mpharrigan — APPROVED** [review]
cool

**Comment by mpharrigan** [comment]
@dstrain115 can you advise on how to fix the type issues

**Comment by dstrain115** [comment]
I think you can add "#type: ignore" to all of them.  The first is because it doesn't know whether it is an attrs class, so it can't attrs.evolve it.

The others are because qualtran and cirq disagree on the definition of "controlled" and the classes inherit from both.   The current order should prefer Qualtran's definition to cirq's which seems fine to me.

**Comment by anurudhp** [comment]
> The others are because qualtran and cirq disagree on the definition of "controlled" and the classes inherit from both. The current order should prefer Qualtran's definition to cirq's which seems fine to me.

I added `type: ignore` to the class definition line. It's strange that the issue does not show up in `GateWithRegisters`, but only here. For example:
```py
class SpecializedSingleQubitControlledGate(GateWithRegisters):
  ...

class QubitizationWalkOperator(SpecializedSingleQubitControlledGate, GateWithRegisters): # <- type check fails here
  ...
```

**Comment by dstrain115** [comment]
I'm not totally sure.  It's something with the multiple inheritance and the function definitions not matching.

File: qualtran/_infra/gate_with_registers.py [hunk 8]
```diff
@@ -548,3 +545,62 @@ def _circuit_diagram_info_(self, args: cirq.CircuitDiagramInfoArgs) -> cirq.Circ
...
+    """Add a specialized single-qubit controlled version of a Bloq.
+
+    `control_val` is an optional single-bit control. When `control_val` is provided,
+     the `control_registers` property should return a single named qubit register,
+     and otherwise return an empty tuple.
+
+    Example usage:
+
+        @attrs.frozen
+        class MyGate(SpecializedSingleQubitControlledGate, GateWithRegisters):
```
[the remark this answers is not in the document]
Comment (anurudhp) [reply qualtran/_infra/gate_with_registers.py]: fixed - used the second suggestion

**Pull request merged**